File: NumericalAnalysis/Python-Implementation/Solution of Equation in One Variable/fix_point_iteration.py
# Fixed-Point Iteration

# INPUT initial approximation P1; tolerance TOL; maximum number of iterations N.
# OUTPUT approximate solution p or message or failure.

# Step 1 Set i = 1.
# Step 2 While i < N do Step 3-6
# 	Step 3 Set p = g(P0). (Compute P).
# 	Step 4 If |p-p0| < TOL then
# 		OUTPUT (p); (The procedure was successful.)
# 		STOP.
# 	Step 5. Set i = i + 1
# 	Step 6. Set P0 = P1. (Update P0).
#   Step 7. OUTPUT ('The method failed after N0 iterations, N =', N0);
# 	    (The procedure was unsuccessful)
# 	    STOP.

File: NumericalAnalysis/Octave-Implementation/Solution of Equation in One Variable/default_template/fix_point_iteration.m
% BSCS 4A
% Borrero, Jan Lance A.
% Cardiño, Joemar J.
% Roque, Kyle Arteal B.
% Zaragoza, Matthew A.

% The main function of our program
function main()

    clc; clear;

    % Main Area for Input
    g = @(x) x^3 + 4 * x^2 - 10; % equation

    p = 1.5; % initial approximation
    TOL = 10^-4; % tolerance 1e-4 or 10 ^ -4
    N = 13; % max num of iterations


    % Function calling
    fixed_point_iteration(g, p, TOL, N);

end

% The Function implemented for Fix-Point Iteration
function fixed_point_iteration(g, p, TOL, N)

    fprintf('fixed-point method\n\n%-10s %-10s %-10s %-10s %-10s\n', 'n', 'a', 'b', 'p', 'f(p)')


    % Step 1
    i = 1;

    % Step 2
    while i <= N

        % Step 3
        P = g(p);

        %fprintf('%-10d %-10f %-10f %-10f\n', i, abs(P - p), p, P);

        % Step 4
        if abs(P - p) < TOL
            p = P;
            return;
        end
        printf("%d\n", p);


        % Step 5
        i = i + 1;

        % Step 6
        p = P;
    end

    % Step 7
    fprintf('The method failed after %d iterations, N = %d\n', i, N);

end

% actual root must be 1.365230013 for this example
% TODO the printing and explanation plus testing

